PushNotifications.md

```diff
@@ -236,17 +234,93 @@ final message = ably.Message(
+
+#### Notification taps
+
+- To be informed when a notification tap launches the app:
+
+```dart
+ably.Push.notificationEvents.notificationTapLaunchedAppFromTerminated
+    .then((remoteMessage) {
+  if (remoteMessage != null) {
+    print('The app was launched by the user by tapping the notification');
+    print(remoteMessage.data);
+  }
+});
+```
+  
+- To be informed when a notification is tapped by the user whilst the app is in the foreground:
+
+```dart
+ably.Push.notificationEvents.onNotificationTap.listen((remoteMessage) {
+  print('Notification was tapped: $remoteMessage');
+});
+```
 
 #### Data Message / Background Notification
 
-ably-flutter currently does not pass the messages to the Flutter application/ Dart-side, so users will need to listen to messages in each platform. See [Implement Push Notifications listener](https://github.com/ably/ably-flutter/issues/141) for more information.
+- On iOS, if a notification is also present, it will be shown before `onMessage` or the callback passed to `ably.Push.notificationEvents.setOnBackgroundMessage` is called.
+- On Android, if a notification is also present, it will be shown after these methods finish (or it's Future completes).
+- When the app is in the foreground, you can listen to messages using:
+
+```dart
+ably.Push.notificationEvents.onMessage.listen((remoteMessage) {
+  print('Message was delivered to app while the app was in the foreground: '
+      '$remoteMessage');
+});
+```
+- This method can be synchronous or `async`.
+
+- When the app is terminated or in the background, you can listen to messages using:
+
+```dart
+Future<void> _backgroundMessageHandler(ably.RemoteMessage message) async {
+  print('Just received a background message, with:');
+  print('RemoteMessage.Notification: ${message.notification}');
+  print('RemoteMessage.Data: ${message.data}');
+}
+ably.Push.notificationEvents.setOnBackgroundMessage(_backgroundMessageHandler);
+```
+  - This method can be synchronous or `async`.
+
+##### Advanced: native message handling
 
-**Android**: This means implementing [`FirebaseMessageService`](https://firebase.google.com/docs/cloud-messaging/android/receive) and overriding `onMessageReceived` method. You must also declare the Service in your `AndroidManifest.xml`. Once you receive your message, you could create a richer notification, by following [Create a Notification](https://developer.android.com/training/notify-user/build-notification).
+Users can listen to messages in each platform using the native message listeners instead of listening to messages on the Dart side. This is not recommended unless you want to **avoid** using other plugins, such as [awesome_notifications](https://pub.dev/packages/awesome_notifications) and [flutter_local_notifications](https://pub.dev/packages/flutter_local_notifications).
```

Comment:
```suggestion
Users can listen to messages in each platform using the native message listeners instead of Dart listeners. This is not recommended unless you want to **avoid** using other plugins, such as [awesome_notifications](https://pub.dev/packages/awesome_notifications) and [flutter_local_notifications](https://pub.dev/packages/flutter_local_notifications).
```